core-blocks/file/index.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { createBlobURL } from '@wordpress/utils';
+import { createBlock } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import edit from './edit';
+
+export const name = 'core/file';
+
+export const settings = {
+	title: __( 'File' ),
+
+	description: __( 'Link to a file.' ),
+
+	icon: 'media-default',
+
+	category: 'common',
+
+	keywords: [ __( 'document' ), __( 'pdf' ) ],
+
+	attributes: {
+		id: {
+			type: 'number',
+		},
+		href: {
+			type: 'string',
+		},
+		fileName: {
+			type: 'string',
+			source: 'text',
+			selector: 'a:not([download])',
+		},
+		// Differs to the href when the block is configured to link to the attachment page
+		textLinkHref: {
+			type: 'string',
+			source: 'attribute',
+			selector: 'a:not([download])',
+			attribute: 'href',
+		},
+		// e.g. `_blank` when the block is configured to open in a new window
+		textLinkTarget: {
+			type: 'string',
+			source: 'attribute',
+			selector: 'a:not([download])',
+			attribute: 'target',
+		},
+		showDownloadButton: {
+			type: 'boolean',
+			default: true,
+		},
+		downloadButtonText: {
+			type: 'string',
+			source: 'text',
+			selector: 'a[download]',
+			default: __( 'Download' ),
+		},
+	},
+
+	supports: {
+		align: true,
+	},
+
+	transforms: {
+		from: [
+			{
+				type: 'files',
+				isMatch: ( files ) => files.length === 1,
+				transform: ( files ) => {
+					const file = files[ 0 ];
+					const blobURL = createBlobURL( file );
+
+					// File will be uploaded in componentDidMount()
+					return createBlock( 'core/file', {
+						href: blobURL,
+						fileName: file.name,
+						textLinkHref: blobURL,
+					} );
+				},
+			},
+			{
+				type: 'block',
+				blocks: [ 'core/audio' ],
+				transform: ( attributes ) => {
+					return createBlock( 'core/file', {
+						href: attributes.src,
+						fileName: attributes.caption && attributes.caption.join(),
+						textLinkHref: attributes.src,
+						id: attributes.id,
+					} );
+				},
+			},
+			{
+				type: 'block',
+				blocks: [ 'core/video' ],
+				transform: ( attributes ) => {
+					return createBlock( 'core/file', {
+						href: attributes.src,
+						fileName: attributes.caption && attributes.caption.join(),
+						textLinkHref: attributes.src,
+						id: attributes.id,
+					} );
+				},
+			},
+		],
+		to: [
```

Comment:
It's pretty strange to unconditionally offer transforms to audio and video regardless of file type, but I also understand that there's no matching mechanism in transforms (unlike `isMatch` when converting classic content to blocks). We could tackle this later, but just noting that the UX can be a bit weird (offering a transform that will break if the file is, say, a ZIP file).

Reply:
Yep, agree with that. I've created an issue for this - #7625.

The question is whether we're happy to leave it as it is and merge this, or whether we want to remove the transforms temporarily.

Another alternative could be to show a warning on audio and video blocks if the file type is incorrect.

Reply:
I've no strong opinion either way. Leaning slightly to leaving it as is for now because if we fix the bug in a future PR it aligns better with our goal of having the next release be 'feature complete'.